domains/certificates/Query.API/API/CertificateGenerationSignUpService/CertificateGenerationSignUpServiceImpl.cs

```diff
@@ -0,0 +1,76 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using API.CertificateGenerationSignUpService.Clients;
+using API.CertificateGenerationSignUpService.Repositories;
+using API.MasterDataService;
+using static API.CertificateGenerationSignUpService.CreateSignUpResult;
+
+namespace API.CertificateGenerationSignUpService;
+
+internal class CertificateGenerationSignUpServiceImpl : ICertificateGenerationSignUpService
+{
+    private readonly IMeteringPointsClient client;
+    private readonly ICertificateGenerationSignUpRepository repository;
+
+    public CertificateGenerationSignUpServiceImpl(IMeteringPointsClient client, ICertificateGenerationSignUpRepository repository)
+    {
+        this.client = client;
+        this.repository = repository;
+    }
+
+    public async Task<CreateSignUpResult> Create(string gsrn, string meteringPointOwner, DateTimeOffset startDate, CancellationToken cancellationToken)
+    {
+        var meteringPoints = await client.GetMeteringPoints(meteringPointOwner, cancellationToken);
+        var matchingMeteringPoint = meteringPoints?.MeteringPoints.FirstOrDefault(mp => mp.GSRN == gsrn);
+
+        if (matchingMeteringPoint == null)
+        {
+            return new GsrnNotFound();
+        }
+
+        if (matchingMeteringPoint.Type != MeterType.Production)
+        {
+            return new NotProductionMeteringPoint();
+        }
+
+        // Check if GSRN is already signed up
```

Comment:
If these comments are necessary I think this could be refactored to a method with a descriptive name so we can remove the comments

Reply:
Deleted comments